python/ray/data/dataset.py

```diff
@@ -854,6 +855,115 @@ def aggregate(self, *aggs: AggregateFn) -> U:
+                               or all(isinstance(on_, Callable)
+                                      for on_ in on))))):
+            raise TypeError(
+                f"`on` must be of type {AggregateOnT}, but got {type(on)}")
+
+        if isinstance(on, list) and len(on) == 0:
+            raise ValueError(
+                "When giving a list for `on`, it must be nonempty.")
+
+        try:
+            is_arrow = self._is_arrow_dataset()
+        except ValueError:
+            # Dataset is empty/cleared, let downstream ops handle this.
+            return on
+
+        if is_arrow:
+            # This should be cached from the ._is_arrow_dataset() check, so we
+            # don't fetch and we assert that the schema is not None.
+            schema = self.schema(fetch_if_missing=False)
+            assert schema is not None
+            if len(schema.names) == 0:
+                # Empty dataset, don't validate `on` since we generically
+                # handle empty datasets downstream.
+                return on
+
+            if on is None:
+                # If a null `on` is given for an Arrow Dataset, coerce it to
+                # all columns sans any that we want to skip.
+                if skip_cols is None:
+                    skip_cols = []
+                elif not isinstance(skip_cols, list):
+                    skip_cols = [skip_cols]
+                on = [col for col in schema.names if col not in skip_cols]
+            # Check that column names refer to valid columns.
+            elif isinstance(on, str) and on not in schema.names:
+                raise ValueError(
+                    f"on={on} is not a valid column name: {schema.names}")
+            elif isinstance(on, list) and isinstance(on[0], str):
+                for on_ in on:
+                    if on_ not in schema.names:
+                        raise ValueError(
+                            f"on={on_} is not a valid column name: "
+                            f"{schema.names}")
+        else:
+            if isinstance(on, str) or (isinstance(on, list)
+                                       and isinstance(on[0], str)):
+                raise ValueError(
+                    "Can't aggregate on a column when using a simple Dataset; "
+                    "use a callable `on` argument or use an Arrow Dataset "
+                    "instead of a simple Dataset.")
+        return on
+
+    def _is_arrow_dataset(self) -> bool:
+        """Determine if Dataset is an Arrow dataset.
+
+        This may block; if the schema is unknown, this will synchronously fetch
+        the schema for the first block.
+        """
+        try:
+            import pyarrow as pa
+        except ModuleNotFoundError:
+            return False
+        else:
+            # We need schema to properly validate, so synchronously
+            # fetch it if necessary.
+            schema = self.schema(fetch_if_missing=True)
+            if schema is None:
```

Comment:
Even if schema might be None, if we check the type of `Block`, we should always know if a dataset is arrow or not? The old `_check_is_arrow` was doing that.

Also for my own learning: is reading parquet the only case where we may not know the schema and in this case we know it must be arrow dataset?

Reply:
This case is only encountered if the Dataset is empty, i.e. there are no blocks, so there's no block to check, which is why the old `_check_is_arrow` wouldn't work here.

>Also for my own learning: is reading parquet the only case where we may not know the schema and in this case we know it must be arrow dataset?

Actually, we're guaranteed to know the schema when reading Parquet; it's the file-based data source with the strongest such guarantee. With the current datasource logic, the only case in which we won't know the schema is if the dataset is empty or has been cleared.